File: src/banking/pages/customer_dashboard_page.py
"""Customer dashboard page object for XYZ Banking Project."""
import re

from playwright.sync_api import Page

from src.banking.helpers.browser_helpers import wait_for_angular
from src.banking.locators.selectors import CustomerDashboard


class CustomerDashboardPage:
    def __init__(self, page: Page):
        self.page = page

    def get_welcome_text(self) -> str:
        return (self.page.locator(CustomerDashboard.welcome_name).text_content() or "").strip()

    def get_account_number(self) -> str:
        return (self.page.locator(CustomerDashboard.account_number).text_content() or "").strip()

    def get_balance(self) -> int:
        text = self.page.locator(CustomerDashboard.balance).text_content() or "0"
        return int(re.sub(r"[^0-9.-]", "", text) or "0")

    def get_currency(self) -> str:
        return (self.page.locator(CustomerDashboard.currency).text_content() or "").strip()

    def get_account_options(self) -> list[str]:
        raw = self.page.locator(f"{CustomerDashboard.account_select} option").all_text_contents()
        return [s.strip() for s in raw if s and s.strip()]

    def get_selected_account(self) -> str:
        """Get selected account, stripping AngularJS prefix like 'number:1004' or '1004 : 1001'."""
        raw = self.page.locator(CustomerDashboard.account_select).input_value()
        cleaned = re.sub(r"^[a-z]+:", "", raw, flags=re.I).strip()
        if " : " in cleaned:
            return cleaned.split(" : ")[0].strip()
        return cleaned or raw

    def select_account(self, account_number: str) -> None:
        loc = self.page.locator(CustomerDashboard.account_select)
        try:
            loc.select_option(value=account_number)
        except Exception:
            loc.select_option(label=account_number)
        wait_for_angular(self.page)

    def click_logout(self) -> None:
        self.page.locator(CustomerDashboard.logout_btn).click()
        wait_for_angular(self.page)

    def expect_loaded(self) -> None:
        self.page.locator(CustomerDashboard.welcome_name).wait_for(state="visible", timeout=15000)

File: playwright-ui/pages/CustomerDashboardPage.ts
/**
 * Customer dashboard page object for XYZ Banking Project.
 * Mirrors src/banking/pages/customer_dashboard_page.py.
 */
import type { Page } from '@playwright/test';
import { waitForAngular } from './helpers';
import { CustomerDashboard } from '../locators/selectors';

export class CustomerDashboardPage {
  constructor(private readonly page: Page) {}

  async getWelcomeText(): Promise<string> {
    return (await this.page.locator(CustomerDashboard.welcome_name).textContent())?.trim() ?? '';
  }

  async getBalance(): Promise<number> {
    const text = (await this.page.locator(CustomerDashboard.balance).textContent()) ?? '0';
    const digits = text.replace(/[^0-9.-]/g, '') || '0';
    return parseInt(digits, 10);
  }

  async getAccountOptions(): Promise<string[]> {
    const raw = await this.page.locator(`${CustomerDashboard.account_select} option`).allTextContents();
    return raw.map((s) => s.trim()).filter(Boolean);
  }

  /** Wait for dropdown to be ready. */
  private static readonly WAIT_TIMEOUT = 10_000;
  /** Per-attempt timeout: keep total under test timeout so we don't get "page closed". */
  private static readonly SELECT_ATTEMPT_TIMEOUT = 6_000;

  async selectAccount(accountNumberOrLabel: string): Promise<void> {
    const loc = this.page.locator(CustomerDashboard.account_select);
    await loc.waitFor({ state: 'visible', timeout: CustomerDashboardPage.WAIT_TIMEOUT });
    const opts = { timeout: CustomerDashboardPage.SELECT_ATTEMPT_TIMEOUT };
    const attempts = [
      () => loc.selectOption({ value: accountNumberOrLabel }, opts),
      () => loc.selectOption({ value: `number:${accountNumberOrLabel}` }, opts),
      () => loc.selectOption({ label: accountNumberOrLabel }, opts),
      () => this.selectAccountByMatchingOption(loc, accountNumberOrLabel, opts),
    ];
    for (const attempt of attempts) {
      try {
        await attempt();
        await waitForAngular(this.page);
        return;
      } catch {
        // try next
      }
    }
    await loc.selectOption({ label: accountNumberOrLabel }, opts);
    await waitForAngular(this.page);
  }

  /** Fallback: select option whose value or text contains the account number. */
  private async selectAccountByMatchingOption(
    selectLoc: ReturnType<Page['locator']>,
    accountNumber: string,
    opts: { timeout: number }
  ): Promise<void> {
    const options = await selectLoc.locator('option').all();
    for (const opt of options) {
      const value = (await opt.getAttribute('value')) ?? '';
      const text = (await opt.textContent()) ?? '';
      if (value.includes(accountNumber) || text.includes(accountNumber)) {
        await selectLoc.selectOption({ value }, opts);
        return;
      }
    }
    throw new Error(`No option found for account ${accountNumber}`);
  }

  async clickLogout(): Promise<void> {
    await this.page.locator(CustomerDashboard.logout_btn).click();
    await waitForAngular(this.page);
  }

  async expectLoaded(): Promise<void> {
    await this.page.locator(CustomerDashboard.welcome_name).waitFor({ state: 'visible', timeout: 15000 });
  }
}
